dali/operators/math/expressions/arithmetic.h

```diff
@@ -155,58 +172,58 @@ inline std::vector<ExprImplTask> CreateExecutionTasks(const ExprNode &expr, Expr
-  const TensorListShape<> *out_shape = nullptr;
-  bool only_scalars = true;
-  for (int i = 0; i < shapes.size(); i++) {
-    bool scalar_input = IsScalarLike(*shapes[i]);
-    if (only_scalars) {
-      if (!out_shape || shapes[i]->sample_dim() > out_shape->sample_dim() || !scalar_input)
-        out_shape = shapes[i];
-      if (!scalar_input)
-        only_scalars = false;
-    } else {
-      DALI_ENFORCE(scalar_input || *out_shape == *shapes[i],
-                   make_string("Input shapes of elemenetwise arithemtic operator \"", op,
-                               "\" do not match. Expected equal shapes, got: ", op, "(",
-                               *out_shape, ", ", *shapes[i], ")."));
-    }
-  }
-  return *out_shape;
-}
-
 /**
  * @brief Recurse over expression tree and propagate shapes between expression nodes.
  *
- * @return The resulting shape of the expression
+ * @return False if the operand shapes required no broadcasting, by being either
+ *         identical shapes or scalar-like. True if the shapes needed broadcasting.
  */
 template <typename Backend>
-DLL_PUBLIC inline const TensorListShape<> &PropagateShapes(ExprNode &expr,
-                                                           const workspace_t<Backend> &ws,
-                                                           int batch_size) {
+DLL_PUBLIC inline bool PropagateShapes(TensorListShape<> &result_shape,
+                                       ExprNode &expr,
+                                       const workspace_t<Backend> &ws,
+                                       int batch_size) {
   if (expr.GetNodeType() == NodeType::Constant) {
     expr.SetShape(TensorListShape<0>(batch_size));
-    return expr.GetShape();
+    result_shape = expr.GetShape();
+    return false;
   }
   if (expr.GetNodeType() == NodeType::Tensor) {
     auto &e = dynamic_cast<ExprTensor &>(expr);
     expr.SetShape(ws.template Input<Backend>(e.GetInputIndex()).shape());
-    return expr.GetShape();
+    result_shape = expr.GetShape();
+    return false;
   }
   auto &func = dynamic_cast<ExprFunc &>(expr);
   int subexpression_count = expr.GetSubexpressionCount();
   DALI_ENFORCE(0 < subexpression_count && subexpression_count <= kMaxArity,
                "Only unary, binary and ternary expressions are supported");
 
-  SmallVector<const TensorListShape<> *, kMaxArity> shapes;
+
+  SmallVector<TensorListShape<>, kMaxArity> shapes;
   shapes.resize(subexpression_count);
+  bool needed_broadcasting = false;
   for (int i = 0; i < subexpression_count; i++) {
-    shapes[i] = &PropagateShapes<Backend>(func[i], ws, batch_size);
+    needed_broadcasting |= PropagateShapes<Backend>(shapes[i], func[i], ws, batch_size);
+  }
+  SmallVector<const TensorListShape<>*, kMaxArity> shapes_ptrs;
+  shapes.reserve(shapes.size());
+  for (const auto& sh : shapes)
+    shapes_ptrs.push_back(&sh);
+  auto shapes_ptrs_span = make_span(shapes_ptrs);
+  if (!CanBroadcast(shapes_ptrs_span)) {
```

Comment:
Perhaps it'd be easier to use a `ValidateShapes` that would check and report an error in one go - this would make it easier to point the user to the offending sample / dimension than just dumping all samples.

Reply:
Actually BroadcastShapes already does error validation and produces a better error message. Removing this previous check